### PR TITLE
Add missing device guard in the reader prefetch thread

### PR DESCRIPTION
**Pull request by JanuszL**
- for the GPU reader - like VideoReader there is a need
  to use a device guard in the thread it operates - the
  prefetch one

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/executor/executor.h [hunk 1]
```diff
@@ -412,6 +412,7 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunGPU() {
       TimeRange tr("[Executor] Run GPU op " + op_node.instance_name,
           TimeRange::knvGreen);
       op.Run(&ws);
+      CUDA_CALL(cudaStreamSynchronize(ws.stream()));
```
Comment (jantonguirao) [inline dali/pipeline/executor/executor.h]: why is this required?

File: dali/test/python/test_video_pipeline.py [hunk 2]
```diff
@@ -70,3 +77,10 @@ def test_stride_video_pipeline():
         pipe_out = pipe.run()
     del pipe
 
+def test_multi_gpu_video_pipeline():
+    gpus = get_gpu_num()
+    pipes = [VideoPipe(batch_size=BATCH_SIZE, data=VIDEO_FILES, device_id=d, num_shards=gpus) for d in range(gpus)]
+    for p in pipes:
+        p.build()
+        p.run()
+test_multi_gpu_video_pipeline()
```
Comment (jantonguirao) [inline dali/test/python/test_video_pipeline.py]: shouldn't this be run with nosetests? If so, remove the invocation

File: dali/pipeline/executor/executor.h (hunk 1 above)
Comment (JanuszL) [reply dali/pipeline/executor/executor.h]: Debug leftover

File: dali/test/python/test_video_pipeline.py (hunk 2 above)
Comment (JanuszL) [reply dali/test/python/test_video_pipeline.py]: Debug leftover

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/operators/reader/reader_op.h [hunk 3]
```diff
@@ -61,7 +61,12 @@ class DataReader : public Operator<Backend> {
         curr_batch_producer_(0),
         consumer_cycle_(false),
         producer_cycle_(false),
-        samples_processed_(0) {}
+        device_id_(-1),
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/reader_op.h]: what's the value of `spec.GetArgument<int>("device_id")` in the case of CPU backend? Maybe we can read it always for simplicity. Just a thought

File: dali/test/python/test_video_pipeline.py [hunk 4]
```diff
@@ -34,11 +34,18 @@
 BATCH_SIZE=4
 COUNT=5
 
+def get_gpu_num():
```
Comment (jantonguirao) [inline dali/test/python/test_video_pipeline.py]: maybe move this to `test_utils.py`

Comment (JanuszL) [reply dali/test/python/test_video_pipeline.py]: test_utils.py imports `PIL` and impose dependency on that, and whoever is uses that needs to have `PIL` installed. I can rework test_utils.py to import this package only when given functionality is used.

File: dali/pipeline/operators/reader/reader_op.h (hunk 3 above)
Comment (JanuszL) [reply dali/pipeline/operators/reader/reader_op.h]: The value is valid, no matter if this is GPU or CPU op. But we assume that, at least for the device guard that device_id == -1 is CPU and noop for that. Maybe in the future when we make DALI able to run without GPU device_id == -1 could be used from python as well to tell pipeline that no GPU will be used.

File: dali/test/python/test_video_pipeline.py (hunk 4 above)
Comment (JanuszL) [reply dali/test/python/test_video_pipeline.py]: Done

**Comment by JanuszL** [comment]
!build

**Comment by jantonguirao** [comment]
!build
